pkg/k8s/watchers/cilium_endpoint_slice_subscriber_test.go

```diff
@@ -0,0 +1,497 @@
+	}
+	return "", true
+}
+
+type fakeEndpointCache struct{}
+
+func (fe *fakeEndpointCache) LookupPodName(namespacedName string) *endpoint.Endpoint {
+	return nil
+}
+
+func createCES(name, namespace string, endpoints []v2alpha1.CoreCiliumEndpoint) *v2alpha1.CiliumEndpointSlice {
+	return &v2alpha1.CiliumEndpointSlice{
+		Namespace: namespace,
+		ObjectMeta: v1.ObjectMeta{
+			Name: name,
+		},
+		Endpoints: endpoints,
+	}
+}
+
+func createEndpoint(name, namespace string, id int64) *types.CiliumEndpoint {
+	return &types.CiliumEndpoint{
+		ObjectMeta: slim_metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+		Identity:   &v2.EndpointIdentity{ID: id},
+		Encryption: &v2.EncryptionSpec{},
+	}
+}
+
+// TestCESSubscriber_CEPTransfer tests a CEP being transferred between two
+// CESes. The order of events:
+// 1. CES add event for the new CES with latest CEP state
+// 2. CES add event for the old CES with old CEP state
+// 3. CES delete event for the old CES.
+func TestCESSubscriber_CEPTransferOnStartup(t *testing.T) {
+	cepNamespace := "ns1"
+	cepName := "cep1"
+	newCEPID := int64(3)
+	oldCEPID := int64(2)
+	fakeEPWatcher := createFakeEPWatcher()
+	fakeEndpointCache := &fakeEndpointCache{}
+	cesSub := &cesSubscriber{
+		epWatcher: fakeEPWatcher,
+		epCache:   fakeEndpointCache,
+		cepMap:    newCEPToCESMap(),
+	}
+	// Add for new CES
+	cesSub.OnAdd(
+		createCES("new-ces", cepNamespace, []v2alpha1.CoreCiliumEndpoint{
+			{
+				Name:       cepName,
+				IdentityID: newCEPID,
+			},
+		}))
+	diff, ok := fakeEPWatcher.expectUpdate(endpointUpdate{
+		newEp: createEndpoint("cep1", "ns1", newCEPID),
+	})
+	if !ok {
+		t.Error(diff)
+	}
+	// Add for old CES
+	cesSub.OnAdd(
+		createCES("old-ces", cepNamespace, []v2alpha1.CoreCiliumEndpoint{
+			{
+				Name:       cepName,
+				IdentityID: oldCEPID,
+			},
+		}))
+	diff, ok = fakeEPWatcher.expectUpdate(endpointUpdate{
+		oldEp: createEndpoint("cep1", "ns1", newCEPID),
+		newEp: createEndpoint("cep1", "ns1", oldCEPID),
+	})
+	if !ok {
+		t.Error(diff)
+	}
+	// Delete the old CES
+	cesSub.OnDelete(
+		createCES("old-ces", cepNamespace, []v2alpha1.CoreCiliumEndpoint{
+			{
+				Name:       cepName,
+				IdentityID: oldCEPID,
+			},
+		}))
+	diff, ok = fakeEPWatcher.expectUpdate(endpointUpdate{
+		oldEp: createEndpoint("cep1", "ns1", oldCEPID),
+		newEp: createEndpoint("cep1", "ns1", newCEPID),
+	})
+	if !ok {
+		t.Error(diff)
+	}
+	diff, ok = fakeEPWatcher.expectNoDelete()
+	if !ok {
+		t.Error(diff)
+	}
+	wantCEPMap := map[string]map[string]*types.CiliumEndpoint{
+		"ns1/cep1": {
+			"new-ces": createEndpoint("cep1", "ns1", 3),
+		},
+	}
+	if diff := cmp.Diff(wantCEPMap, cesSub.cepMap.cepMap); diff != "" {
+		t.Errorf("Unexpected CEP map (-want +got):\n%s", diff)
+	}
+}
+
+// TestCESSubscriber_CEPTransferViaUpdate tests a CEP being transferred between
+// two CESes. The order of events:
+// 1. CES update event for the old CES with old CEP state
```

Comment:
add one more step here - add for old CES?

Reply:
Done